### PR TITLE
Add aligned malloc/free handler for Ruby 2.4+ 

### PR DESCRIPTION
**Pull request by Watson1978**
This PR is yet another implementation that solves the same problem as #832.
#832 requires adding new API in Ruby, and it might not be merged.

So, this implements custom allocation handler using `rb_gc_adjust_memory_usage()` Ruby's API.
The API notify memory usage in RMagick/ImageMagick to Ruby GC and GC works properly. 

`rb_gc_adjust_memory_usage()` can be used since Ruby 2.4.
So this PR will support many platform than #832.

### CODE REVIEW

File: ext/RMagick/rmmain.c [hunk 1]
```diff
@@ -14,6 +14,13 @@
 #define MAIN                        // Define external variables
 #include "rmagick.h"
 
+#if defined HAVE_SETMAGICKALIGNEDMEMORYMETHODS
+#if defined HAVE_RB_GC_ADJUST_MEMORY_USAGE && \
+    (defined HAVE_POSIX_MEMALIGN || defined HAVE_MEMALIGN || defined _WIN32)
```
Comment (mockdeep) [inline ext/RMagick/rmmain.c]: @Watson1978 what platforms do and don't have these options? Is it just Ruby 2.3 that doesn't have them, or are there others? Will we be able to remove these checks at some point?

Comment (Watson1978) [reply ext/RMagick/rmmain.c]: The platforms that posix API cannot be used ￼and they doesn’t have ImageMagick 7.0.9

Comment (Watson1978) [reply ext/RMagick/rmmain.c]: I have tried with macOS, Linux and Windows 10. So, almost platforms will be supported, I think.

**Comment by mockdeep** [comment]
@dlemstra do you mind taking a look at this?

**Comment by Watson1978** [comment]
Seems that ImageMagick 6.9.10-79 also has implemented `SetMagickAlignedMemoryMethods()` which is same with ImageMagick 7 and it uses huge memory without this PR.

File: ext/RMagick/rmmain.c [hunk 2]
```diff
@@ -98,6 +110,75 @@ static void rm_free(void *ptr)
...
+/**
+ * Free aligned memory.
+ *
+ * No Ruby usage (internal function)
+ *
+ * @param ptr pointer to the existing block of memory
+ */
+static void rm_aligned_free(void *ptr)
+{
+    size_t size = rm_aligned_malloc_size(ptr);
```
Comment (dlemstra) [inline ext/RMagick/rmmain.c]: It looks like we would also need `alignment` here but I don't have an idea where that should be stored then. We have a fixed value in our header file:

`memory-private.h:`
```
#if defined(__powerpc__)
#  define CACHE_LINE_SIZE  (16 * MAGICKCORE_SIZEOF_VOID_P)
#else
#  define CACHE_LINE_SIZE  (8 * MAGICKCORE_SIZEOF_VOID_P)
#endif
```

```
  if (memory_methods.acquire_aligned_memory_handler != (AcquireAlignedMemoryHandler) NULL)
    return(memory_methods.acquire_aligned_memory_handler(size,CACHE_LINE_SIZE));
```

So maybe we could just use that instead?

File: ext/RMagick/rmmain.c [hunk 3]
```diff
@@ -98,6 +110,75 @@ static void rm_free(void *ptr)
...
+    if (posix_memalign(&res, alignment, size) != 0) {
+        return NULL;
+    }
+#elif defined(HAVE_MEMALIGN)
+    res = memalign(alignment, size);
+#elif defined(_WIN32)
+    res = _aligned_malloc(size, alignment);
+#endif
+
+    size = rm_aligned_malloc_size(res);
```
[the remark this answers is not in the document]
Comment (Watson1978) [reply ext/RMagick/rmmain.c]: Fixed with https://github.com/rmagick/rmagick/pull/836/commits/74e4b5be562d6dd84b518ac4e14968ba913aca78

File: ext/RMagick/rmmain.c [hunk 4]
```diff
@@ -98,6 +110,75 @@ static void rm_free(void *ptr)
...
+
+static size_t
+rm_aligned_malloc_size(void *ptr)
+{
+#if defined(HAVE_MALLOC_USABLE_SIZE)
+    return malloc_usable_size(ptr);
+#elif defined(HAVE_MALLOC_SIZE)
+    return malloc_size(ptr);
+#elif defined(_WIN32)
+    return _aligned_msize(ptr, 0, 0);
```
[the remark this answers is not in the document]
Comment (Watson1978) [reply ext/RMagick/rmmain.c]: https://github.com/rmagick/rmagick/pull/836/commits/8f0ef8350c1435988b9cd576aca50ce4e15b4115

File: ext/RMagick/rmmain.c (hunk 2 above)
Comment (Watson1978) [reply ext/RMagick/rmmain.c]: https://github.com/rmagick/rmagick/pull/836/commits/8f0ef8350c1435988b9cd576aca50ce4e15b4115

**Comment by Watson1978** [comment]
Thanks 🎉

**Pull request merged**